File: mac/AnimationConfig.m
#import "AnimationConfig.h"

@implementation AnimationConfig {
    Animation _animations[MAX_TOTALANIMATIONS];
    BOOL _fixedLegs;
    BOOL _fixedTorso;
}

- (const Animation *)animations {
    return _animations;
}

- (instancetype)initWithData:(NSData *)data {
    self = [super init];
    if (!self) return nil;

    memset(_animations, 0, sizeof(_animations));
    _fixedLegs = NO;
    _fixedTorso = NO;

    NSString *text = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
    if (!text) {
        text = [[NSString alloc] initWithData:data encoding:NSISOLatin1StringEncoding];
    }
    if (!text) return nil;

    const char *cstr = [text UTF8String];
    const char *p = cstr;
    int skip = 0;

    // Skip optional header keywords
    while (*p) {
        // Skip whitespace and newlines
        while (*p && (*p == ' ' || *p == '\t' || *p == '\r' || *p == '\n')) p++;
        if (!*p) break;

        // Skip comment lines (// comments)
        if (p[0] == '/' && p[1] == '/') {
            while (*p && *p != '\n') p++;
            continue;
        }

        // If it starts with a digit or minus, we've reached the animation data
        if ((*p >= '0' && *p <= '9') || *p == '-') break;

        // Parse known keywords
        char token[256];
        int ti = 0;
        while (*p && *p != ' ' && *p != '\t' && *p != '\r' && *p != '\n' && ti < 255) {
            token[ti++] = *p++;
        }
        token[ti] = '\0';

        if (strcasecmp(token, "footsteps") == 0) {
            // Skip the footstep type
            while (*p && (*p == ' ' || *p == '\t')) p++;
            while (*p && *p != ' ' && *p != '\t' && *p != '\r' && *p != '\n') p++;
        } else if (strcasecmp(token, "headoffset") == 0) {
            // Skip 3 floats
            for (int i = 0; i < 3; i++) {
                while (*p && (*p == ' ' || *p == '\t')) p++;
                while (*p && *p != ' ' && *p != '\t' && *p != '\r' && *p != '\n') p++;
            }
        } else if (strcasecmp(token, "sex") == 0) {
            while (*p && (*p == ' ' || *p == '\t')) p++;
            while (*p && *p != ' ' && *p != '\t' && *p != '\r' && *p != '\n') p++;
        } else if (strcasecmp(token, "fixedlegs") == 0) {
            _fixedLegs = YES;
        } else if (strcasecmp(token, "fixedtorso") == 0) {
            _fixedTorso = YES;
        }
    }

    // Parse animation entries
    for (int i = 0; i < MAX_ANIMATIONS; i++) {
        // Skip whitespace, newlines, comments
        while (*p) {
            while (*p && (*p == ' ' || *p == '\t' || *p == '\r' || *p == '\n')) p++;
            if (p[0] == '/' && p[1] == '/') {
                while (*p && *p != '\n') p++;
                continue;
            }
            break;
        }

        if (!*p) {
            // Handle missing team animations (TORSO_GETFLAG through TORSO_NEGATIVE)
            if (i >= TORSO_GETFLAG && i <= TORSO_NEGATIVE) {
                _animations[i] = _animations[TORSO_GESTURE];
                continue;
            }
            break;
        }

        // firstFrame
        _animations[i].firstFrame = atoi(p);
        while (*p && *p != ' ' && *p != '\t') p++;

        // Compute leg frame offset at LEGS_WALKCR
        if (i == LEGS_WALKCR) {
            skip = _animations[LEGS_WALKCR].firstFrame - _animations[TORSO_GESTURE].firstFrame;
        }
        if (i >= LEGS_WALKCR && i < TORSO_GETFLAG) {
            _animations[i].firstFrame -= skip;
        }

        // numFrames
        while (*p && (*p == ' ' || *p == '\t')) p++;
        int numFrames = atoi(p);
        while (*p && *p != ' ' && *p != '\t') p++;
        _animations[i].reversed = 0;
        _animations[i].flipflop = 0;
        if (numFrames < 0) {
            numFrames = -numFrames;
            _animations[i].reversed = 1;
        }
        _animations[i].numFrames = numFrames;

        // loopFrames
        while (*p && (*p == ' ' || *p == '\t')) p++;
        _animations[i].loopFrames = atoi(p);
        while (*p && *p != ' ' && *p != '\t') p++;

        // fps
        while (*p && (*p == ' ' || *p == '\t')) p++;
        float fps = atof(p);
        while (*p && *p != ' ' && *p != '\t' && *p != '\r' && *p != '\n') p++;
        if (fps == 0) fps = 1;
        _animations[i].frameLerp = (int)(1000.0f / fps);

        // Skip rest of line (comments etc)
        while (*p && *p != '\n') p++;
    }

    // Extra animations
    _animations[LEGS_BACKCR] = _animations[LEGS_WALKCR];
    _animations[LEGS_BACKCR].reversed = 1;
    _animations[LEGS_BACKWALK] = _animations[LEGS_WALK];
    _animations[LEGS_BACKWALK].reversed = 1;

    _animations[FLAG_RUN].firstFrame = 0;
    _animations[FLAG_RUN].numFrames = 16;
    _animations[FLAG_RUN].loopFrames = 16;
    _animations[FLAG_RUN].frameLerp = (int)(1000.0f / 15.0f);
    _animations[FLAG_RUN].reversed = 0;

    _animations[FLAG_STAND].firstFrame = 16;
    _animations[FLAG_STAND].numFrames = 5;
    _animations[FLAG_STAND].loopFrames = 0;
    _animations[FLAG_STAND].frameLerp = (int)(1000.0f / 20.0f);
    _animations[FLAG_STAND].reversed = 0;

    _animations[FLAG_STAND2RUN].firstFrame = 16;
    _animations[FLAG_STAND2RUN].numFrames = 5;
    _animations[FLAG_STAND2RUN].loopFrames = 1;
    _animations[FLAG_STAND2RUN].frameLerp = (int)(1000.0f / 15.0f);
    _animations[FLAG_STAND2RUN].reversed = 1;

    return self;
}

@end

File: win/MD3View/AnimationConfig.cs
using System.Text;

namespace MD3View;

public class AnimationConfig
{
    private readonly Animation[] _animations = new Animation[(int)AnimNumber.MAX_TOTALANIMATIONS];

    public bool FixedLegs { get; private set; }
    public bool FixedTorso { get; private set; }

    public Animation[] Animations => _animations;

    public AnimationConfig(byte[] data)
    {
        string? text;
        try
        {
            text = Encoding.UTF8.GetString(data);
        }
        catch
        {
            text = Encoding.Latin1.GetString(data);
        }

        if (text == null) return;

        Parse(text);
    }

    private void Parse(string text)
    {
        int pos = 0;
        int skip = 0;

        // Skip optional header keywords
        while (pos < text.Length)
        {
            SkipWhitespace(text, ref pos);
            if (pos >= text.Length) break;

            // Skip comment lines
            if (pos + 1 < text.Length && text[pos] == '/' && text[pos + 1] == '/')
            {
                SkipToEndOfLine(text, ref pos);
                continue;
            }

            // If starts with digit or minus, we've reached animation data
            if (char.IsDigit(text[pos]) || text[pos] == '-') break;

            // Parse known keywords
            var token = ReadToken(text, ref pos);

            if (string.Equals(token, "footsteps", StringComparison.OrdinalIgnoreCase))
            {
                SkipSpaces(text, ref pos);
                SkipToken(text, ref pos);
            }
            else if (string.Equals(token, "headoffset", StringComparison.OrdinalIgnoreCase))
            {
                for (int i = 0; i < 3; i++)
                {
                    SkipSpaces(text, ref pos);
                    SkipToken(text, ref pos);
                }
            }
            else if (string.Equals(token, "sex", StringComparison.OrdinalIgnoreCase))
            {
                SkipSpaces(text, ref pos);
                SkipToken(text, ref pos);
            }
            else if (string.Equals(token, "fixedlegs", StringComparison.OrdinalIgnoreCase))
            {
                FixedLegs = true;
            }
            else if (string.Equals(token, "fixedtorso", StringComparison.OrdinalIgnoreCase))
            {
                FixedTorso = true;
            }
        }

        // Parse animation entries
        for (int i = 0; i < (int)AnimNumber.MAX_ANIMATIONS; i++)
        {
            // Skip whitespace, newlines, comments
            while (pos < text.Length)
            {
                SkipWhitespace(text, ref pos);
                if (pos + 1 < text.Length && text[pos] == '/' && text[pos + 1] == '/')
                {
                    SkipToEndOfLine(text, ref pos);
                    continue;
                }
                break;
            }

            if (pos >= text.Length)
            {
                // Handle missing team animations
                if (i >= (int)AnimNumber.TORSO_GETFLAG && i <= (int)AnimNumber.TORSO_NEGATIVE)
                {
                    _animations[i] = _animations[(int)AnimNumber.TORSO_GESTURE];
                    continue;
                }
                break;
            }

            // firstFrame
            _animations[i].FirstFrame = ReadInt(text, ref pos);

            // Compute leg frame offset at LEGS_WALKCR
            if (i == (int)AnimNumber.LEGS_WALKCR)
            {
                skip = _animations[(int)AnimNumber.LEGS_WALKCR].FirstFrame -
                       _animations[(int)AnimNumber.TORSO_GESTURE].FirstFrame;
            }
            if (i >= (int)AnimNumber.LEGS_WALKCR && i < (int)AnimNumber.TORSO_GETFLAG)
            {
                _animations[i].FirstFrame -= skip;
            }

            // numFrames
            SkipSpaces(text, ref pos);
            int numFrames = ReadInt(text, ref pos);
            _animations[i].Reversed = 0;
            _animations[i].Flipflop = 0;
            if (numFrames < 0)
            {
                numFrames = -numFrames;
                _animations[i].Reversed = 1;
            }
            _animations[i].NumFrames = numFrames;

            // loopFrames
            SkipSpaces(text, ref pos);
            _animations[i].LoopFrames = ReadInt(text, ref pos);

            // fps
            SkipSpaces(text, ref pos);
            float fps = ReadFloat(text, ref pos);
            if (fps == 0) fps = 1;
            _animations[i].FrameLerp = (int)(1000.0f / fps);

            // Skip rest of line
            SkipToEndOfLine(text, ref pos);
        }

        // Extra animations
        _animations[(int)AnimNumber.LEGS_BACKCR] = _animations[(int)AnimNumber.LEGS_WALKCR];
        _animations[(int)AnimNumber.LEGS_BACKCR].Reversed = 1;
        _animations[(int)AnimNumber.LEGS_BACKWALK] = _animations[(int)AnimNumber.LEGS_WALK];
        _animations[(int)AnimNumber.LEGS_BACKWALK].Reversed = 1;

        _animations[(int)AnimNumber.FLAG_RUN].FirstFrame = 0;
        _animations[(int)AnimNumber.FLAG_RUN].NumFrames = 16;
        _animations[(int)AnimNumber.FLAG_RUN].LoopFrames = 16;
        _animations[(int)AnimNumber.FLAG_RUN].FrameLerp = (int)(1000.0f / 15.0f);
        _animations[(int)AnimNumber.FLAG_RUN].Reversed = 0;

        _animations[(int)AnimNumber.FLAG_STAND].FirstFrame = 16;
        _animations[(int)AnimNumber.FLAG_STAND].NumFrames = 5;
        _animations[(int)AnimNumber.FLAG_STAND].LoopFrames = 0;
        _animations[(int)AnimNumber.FLAG_STAND].FrameLerp = (int)(1000.0f / 20.0f);
        _animations[(int)AnimNumber.FLAG_STAND].Reversed = 0;

        _animations[(int)AnimNumber.FLAG_STAND2RUN].FirstFrame = 16;
        _animations[(int)AnimNumber.FLAG_STAND2RUN].NumFrames = 5;
        _animations[(int)AnimNumber.FLAG_STAND2RUN].LoopFrames = 1;
        _animations[(int)AnimNumber.FLAG_STAND2RUN].FrameLerp = (int)(1000.0f / 15.0f);
        _animations[(int)AnimNumber.FLAG_STAND2RUN].Reversed = 1;
    }

    private static void SkipWhitespace(string text, ref int pos)
    {
        while (pos < text.Length && (text[pos] == ' ' || text[pos] == '\t' || text[pos] == '\r' || text[pos] == '\n'))
            pos++;
    }

    private static void SkipSpaces(string text, ref int pos)
    {
        while (pos < text.Length && (text[pos] == ' ' || text[pos] == '\t'))
            pos++;
    }

    private static void SkipToEndOfLine(string text, ref int pos)
    {
        while (pos < text.Length && text[pos] != '\n')
            pos++;
    }

    private static void SkipToken(string text, ref int pos)
    {
        while (pos < text.Length && text[pos] != ' ' && text[pos] != '\t' && text[pos] != '\r' && text[pos] != '\n')
            pos++;
    }

    private static string ReadToken(string text, ref int pos)
    {
        int start = pos;
        while (pos < text.Length && text[pos] != ' ' && text[pos] != '\t' && text[pos] != '\r' && text[pos] != '\n')
            pos++;
        return text[start..pos];
    }

    private static int ReadInt(string text, ref int pos)
    {
        int start = pos;
        if (pos < text.Length && (text[pos] == '-' || text[pos] == '+')) pos++;
        while (pos < text.Length && char.IsDigit(text[pos])) pos++;
        if (start == pos) return 0;
        return int.TryParse(text[start..pos], out int val) ? val : 0;
    }

    private static float ReadFloat(string text, ref int pos)
    {
        int start = pos;
        if (pos < text.Length && (text[pos] == '-' || text[pos] == '+')) pos++;
        while (pos < text.Length && (char.IsDigit(text[pos]) || text[pos] == '.')) pos++;
        if (start == pos) return 0;
        return float.TryParse(text[start..pos], System.Globalization.NumberStyles.Float,
            System.Globalization.CultureInfo.InvariantCulture, out float val) ? val : 0;
    }
}
